### PR TITLE
PLAT-135: add metrics for consensus phases

### PR DESCRIPTION
**Pull request by ElSacramento**
Add metrics for phase01

### CODE REVIEW

**Comment by ElSacramento** [comment]
```
# HELP insolar_phase01_latency time spent on phase01
# TYPE insolar_phase01_latency histogram
insolar_phase01_latency_bucket{role="light_material",le="0.001"} 0
insolar_phase01_latency_bucket{role="light_material",le="0.01"} 0
insolar_phase01_latency_bucket{role="light_material",le="0.1"} 153
insolar_phase01_latency_bucket{role="light_material",le="1"} 199
insolar_phase01_latency_bucket{role="light_material",le="10"} 199
insolar_phase01_latency_bucket{role="light_material",le="100"} 199
insolar_phase01_latency_bucket{role="light_material",le="1000"} 199
insolar_phase01_latency_bucket{role="light_material",le="5000"} 199
insolar_phase01_latency_bucket{role="light_material",le="10000"} 199
insolar_phase01_latency_bucket{role="light_material",le="20000"} 199
insolar_phase01_latency_bucket{role="light_material",le="+Inf"} 199
insolar_phase01_latency_sum{role="light_material"} 17.433929999999997
insolar_phase01_latency_count{role="light_material"} 199
```

```
# TYPE insolar_phase01_latency histogram
insolar_phase01_latency_bucket{role="virtual",le="0.001"} 0
insolar_phase01_latency_bucket{role="virtual",le="0.01"} 0
insolar_phase01_latency_bucket{role="virtual",le="0.1"} 147
insolar_phase01_latency_bucket{role="virtual",le="1"} 191
insolar_phase01_latency_bucket{role="virtual",le="10"} 192
insolar_phase01_latency_bucket{role="virtual",le="100"} 192
insolar_phase01_latency_bucket{role="virtual",le="1000"} 192
insolar_phase01_latency_bucket{role="virtual",le="5000"} 192
insolar_phase01_latency_bucket{role="virtual",le="10000"} 192
insolar_phase01_latency_bucket{role="virtual",le="20000"} 192
insolar_phase01_latency_bucket{role="virtual",le="+Inf"} 192
insolar_phase01_latency_sum{role="virtual"} 17.890970999999997
insolar_phase01_latency_count{role="virtual"} 192
```

```
# HELP insolar_phase01_latency time spent on phase01
# TYPE insolar_phase01_latency histogram
insolar_phase01_latency_bucket{role="heavy_material",le="0.001"} 0
insolar_phase01_latency_bucket{role="heavy_material",le="0.01"} 0
insolar_phase01_latency_bucket{role="heavy_material",le="0.1"} 154
insolar_phase01_latency_bucket{role="heavy_material",le="1"} 195
insolar_phase01_latency_bucket{role="heavy_material",le="10"} 195
insolar_phase01_latency_bucket{role="heavy_material",le="100"} 195
insolar_phase01_latency_bucket{role="heavy_material",le="1000"} 195
insolar_phase01_latency_bucket{role="heavy_material",le="5000"} 195
insolar_phase01_latency_bucket{role="heavy_material",le="10000"} 195
insolar_phase01_latency_bucket{role="heavy_material",le="20000"} 195
insolar_phase01_latency_bucket{role="heavy_material",le="+Inf"} 195
insolar_phase01_latency_sum{role="heavy_material"} 16.656591
insolar_phase01_latency_count{role="heavy_material"} 195
```

File: ledger-core/v2/network/consensus/gcpv2/phasebundle/metrics/metrics.go [hunk 1]
```diff
@@ -0,0 +1,55 @@
+// Copyright 2020 Insolar Network Ltd.
```
Comment (sexyhamster) [inline ledger-core/v2/network/consensus/gcpv2/phasebundle/metrics/metrics.go]: move this package to network/concensus, so all concensus metrics will be in one place

Comment (ElSacramento) [reply ledger-core/v2/network/consensus/gcpv2/phasebundle/metrics/metrics.go]: done

File: ledger-core/v2/network/consensus/gcpv2/phasebundle/ph01ctl/phase01.go [hunk 2]
```diff
@@ -160,9 +163,11 @@ func (c *Phase01Controller) workerPhase01(ctx context.Context) {
 
 	var nsh proofs.NodeStateHash
 	startIndex := 0
+	phase01StartedAt := time.Now()
 
 	if ok, nshChannel := c.R.PreparePulseChange(); ok {
 		nsh, startIndex = c.workerSendPhase0(ctx, nodes, nshChannel)
+		stats.Record(ctx, metrics.Phase01Time.M(float64(time.Since(phase01StartedAt).Nanoseconds())/1e6))
```
Comment (cyraxred) [inline ledger-core/v2/network/consensus/gcpv2/phasebundle/ph01ctl/phase01.go:170]: 1. please use named constants
2. this is time critical code. Is there a guarantee that stats.Record will have low latency?

Comment (ruz) [reply ledger-core/v2/network/consensus/gcpv2/phasebundle/ph01ctl/phase01.go:170]: 1. why don't we just measure nanoseconds
2. we want to measure time critical code, measurements take time. what guarantees do we need to say that stats.Record has low latency? Benchmark?

File: ledger-core/v2/network/consensus/gcpv2/phasebundle/ph3ctl/phase3.go [hunk 3]
```diff
@@ -145,6 +148,9 @@ func (c *Phase3Controller) workerPhase3(ctx context.Context) {
 		defer runtime.UnlockOSThread()
 	}
 
+	phase3StartedAt := time.Now()
+	defer stats.Record(ctx, metrics.Phase3Time.M(float64(time.Since(phase3StartedAt).Nanoseconds())/1e6))
```
Comment (ruz) [inline ledger-core/v2/network/consensus/gcpv2/phasebundle/ph3ctl/phase3.go:152]: this is incorrect as deferred call's arguments are evaluated immediately

File: ledger-core/v2/network/consensus/gcpv2/phasebundle/ph01ctl/phase01.go (hunk 2 above)
Comment (ElSacramento) [reply ledger-core/v2/network/consensus/gcpv2/phasebundle/ph01ctl/phase01.go]: done

File: ledger-core/v2/network/consensus/gcpv2/phasebundle/ph3ctl/phase3.go (hunk 3 above)
Comment (ElSacramento) [reply ledger-core/v2/network/consensus/gcpv2/phasebundle/ph3ctl/phase3.go]: done

File: ledger-core/v2/network/consensus/gcpv2/phasebundle/ph3ctl/phase3.go [hunk 4]
```diff
@@ -145,12 +148,16 @@ func (c *Phase3Controller) workerPhase3(ctx context.Context) {
...
+	phase3StartedAt := time.Now()
+
 	localInspector := c.workerPrePhase3(ctx)
 	if localInspector == nil {
 
 		// context was stopped in a hard way, we are dead in terms of consensus
 		// TODO should wait for further packets to decide if we need to turn ourselves into suspended state
 		// c.R.StopRoundByTimeout()
+		// TODO: low-latency metrics - https://insolar.atlassian.net/browse/PLAT-217
+		go stats.Record(ctx, metrics.Phase3Time.M(float64(time.Since(phase3StartedAt).Nanoseconds())*metrics.StatUnit))
```
Comment (ruz) [inline ledger-core/v2/network/consensus/gcpv2/phasebundle/ph3ctl/phase3.go:160]: you can get rid of code duplication using `defer func() { go stats.Record(...) }`

Comment (ElSacramento) [reply ledger-core/v2/network/consensus/gcpv2/phasebundle/ph3ctl/phase3.go]: done

**Pull request merged**